scipy/linalg/_linsolve.pyx.in

```diff
@@ -0,0 +1,603 @@
+
+ctypedef fused lapack_cz_t:
+    (float complex)
+    (double complex)
+
+ctypedef fused lapack_sd_t:
+    float
+    double
+
+ctypedef fused lapack_f_t:
+    float
+    (float complex)
+ctypedef fused lapack_d_t:
+    double
+    (double complex)
+
+ctypedef (int, int) (*f_tri_ptr)(lapack_t[:, ::1]) nogil
+ctypedef (bint, bint) (*f_sym_ptr)(lapack_t[:, ::1]) nogil
+ctypedef int (*f_posdiag_ptr)(lapack_t[:, ::1]) nogil
+
+
+# The numpy facilities for type-casting checks are too slow for small sized
+# arrays and eats away the time budget for the checkups. Here we provide a
+# precomputed version of the numpy.can_cast() table.
+
+# output of np.typecodes['All']
+cdef str casting_key = '?bhilqpBHILQPefdgFDGSUVOMm'
+
+#                          f  d  F  D
+casting_table = np.array([[1, 1, 1, 1],  # ?
+                          [1, 1, 1, 1],  # b
+                          [1, 1, 1, 1],  # h
+                          [0, 1, 0, 1],  # i
+                          [0, 1, 0, 1],  # l
+                          [0, 1, 0, 1],  # q
+                          [0, 1, 0, 1],  # p
+                          [1, 1, 1, 1],  # B
+                          [1, 1, 1, 1],  # H
+                          [0, 1, 0, 1],  # I
+                          [0, 1, 0, 1],  # L
+                          [0, 1, 0, 1],  # Q
+                          [0, 1, 0, 1],  # P
+                          [1, 1, 1, 1],  # e
+                          [1, 1, 1, 1],  # f
+                          [0, 1, 0, 1],  # d
+                          [0, 1, 0, 1],  # g
+                          [0, 0, 1, 1],  # F
+                          [0, 0, 0, 1],  # D
+                          [0, 0, 0, 1],  # G
+                          [0, 0, 0, 0],  # S
+                          [0, 0, 0, 0],  # U
+                          [0, 0, 0, 0],  # V
+                          [0, 0, 0, 0],  # O
+                          [0, 0, 0, 0],  # M
+                          [0, 0, 0, 0]],  # m
+                         dtype=bool)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.nonecheck(False)
+cpdef int get_lapack_flavor(str a, str b, str K = casting_key,
+                            const cnp.npy_bool[:, ::1]C = casting_table):
+    cdef int ind
+    cdef size_t rowa = K.find(a)
+    cdef size_t rowb = K.find(b)
+    if rowa == -1 or rowb == -1:
+        raise TypeError('Data type is not known to NumPy.')
+
+    for ind in range(4):
+        if C[rowa, ind] and C[rowb, ind]:
+            return ind
+
+    return -1
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.nonecheck(False)
+cpdef bint array_is_finite(lapack_t[:, ::1]A) nogil:
+    """
+    Runs over the array and checks for infs and nans. Compared
+    to asarray_chkfinite the performance gets better as the
+    array size increases. Starting from n=3 it is ~x2 faster
+    for finite arrays and faster if array has inf/nan due to
+    early exit capability.
+    """
+    cdef int n = <int>A.shape[0]
```

Comment:
This is assumed `size_t` in other places, would prevent overflow if matrices are really large or LinearOperator used

Reply:
Ah yes, I'll fix that.